ziglang/zig
# feat: zig cc support openharmony

This PR is derived from PR [#20020](https://github.com/ziglang/zig/pull/20020), aimed at adding OpenHarmony operating system support for the `zig cc` command. Since OpenHarmony's C standard library is based on musl implementation, we believe that only adding the `available_libcs` is necessary.

Additionally, we have implemented several benchmark tests based on some fundamental capabilities to ensure usability. I will provide the testing steps and my testing conclusions below.

## How to Test?

1. You need to register on the official website and [download](https://developer.huawei.com/consumer/cn/develop/) the latest version of DevEco Studio. After installation, you need to install the emulator to provide the testing environment.
2. Next, you need to set an environment variable to provide platform-specific capabilities for the compiler:
   ```
   export OHOS_NDK_HOME=/Applications/DevEco-Studio.app/Contents/sdk/default/openharmony
   export PATH=$PATH:$OHOS_NDK_HOME/native/build-tools/cmake/bin:/Applications/DevEco-Studio.app/Contents/tools/ohpm/bin
   export PATH=$OHOS_NDK_HOME/toolchains:$OHOS_NDK_HOME/native/llvm/bin:$PATH
   ```
3. Before starting the tests, we need to install some tools: [cargo-zigbuild](https://github.com/rust-cross/cargo-zigbuild)  [just](https://github.com/casey/just)  [Rust](https://www.rust-lang.org/) and install some toolchains.
    ```bash
   rustup target add aarch64-unknown-linux-ohos
   rustup target add armv7-unknown-linux-ohos
   rustup target add x86_64-unknown-linux-ohos
    ```

4. Clone the [ohos-rs](https://github.com/ohos-rs/ohos-rs) repository and proceed with the build
    > Note: branch should be `ohos`   
     
   Build example as first and just run command with below:
     ```bash
     just zig-build
     ```

6. Open the `harmony-example` project from ohos-rs in the IDE, then right-click to run the unit tests as shown
  
     <img width="519" alt="image" src="https://github.com/user-attachments/assets/be888891-11e4-4af6-ade3-dcd13babdc18">

7. Run the complete test suite via command line to view the results
    ```bash
    hdc shell aa test -b com.ohos.napi -m entry_test -s unittest /ets/testrunner/OpenHarmonyTestRunner
   ```

## Result

Based on my current observations, the test results are consistent with those built using the NDK. All unit test cases are sourced from [napi-rs](https://github.com/napi-rs/napi-rs). While some test cases haven't passed due to system implementation issues, I believe these can be disregarded. The successfully passed tests are sufficient to confirm the functionality of using `zig cc` as the compiler
